src/java/org/apache/cassandra/db/commitlog/SimpleCachedBufferPool.java

```diff
@@ -79,19 +91,20 @@ public ByteBuffer createBuffer(BufferType bufferType)
 
     public ByteBuffer getThreadLocalReusableBuffer(int size)
     {
-        ByteBuffer result = reusableBufferHolder.get();
-        if (result.capacity() < size || BufferType.typeOf(result) != preferredReusableBufferType)
+        ByteBuffer result = reusableBB.get();
+        if (result.capacity() < size)
         {
             FileUtils.clean(result);
             result = preferredReusableBufferType.allocate(size);
-            reusableBufferHolder.set(result);
+            reusableBB.set(result);
         }
         return result;
     }
 
     public void setPreferredReusableBufferType(BufferType type)
     {
         preferredReusableBufferType = type;
+        reusableBB = reusableBBHolder.get(preferredReusableBufferType);
```

Comment:
Should we not have `preferredReusableBufferType` and  `reusableBB` be `volatile` variables and the   `setPreferredReusableBufferType` method `synchronized` ? With the current implementation, it seems to me that, if a thread call `setPreferredReusableBufferType` and another `getThreadLocalReusableBuffer` the output of `getThreadLocalReusableBuffer` is unpredictable and the wrong buffer type can be added the `reusableBB` field.
Am I missing something?

Reply:
Making them volatile won't change much, the right thing to do is to not allow the preferred type to change. Is it possible to pass the type as a constructor argument and make `reuseableBB` final?

Reply:
You are right we can still have a race.
The preferedType is set in the `CompressedSegment` and int the `EncryptedSegment` constructor. I do not recall if we can change the commit log configuration without restarting the server. If not we can figure out which type will be the prefered one based on the `commitLog.configuration`. It is not super clean has it some how duplicate the logic from `createSegment` but it should work.

Reply:
Given this `getThreadLocalReusableBuffer` seems to be isolated within the class, as it is not tied in with the buffer pool functinality, I don't like adding this param to the constructor. It reads as if this was the preferred type to the BBs in the pool... sounds confusing to me, changes the 'main' API, it might not get used if you only use the pool functionality

I would like to propose the latest commit which does away with both variables. The price to pay is the Map lookup. I read a bit about `EnumMap`'s implementation and it is an Array indexed by the Enum int. So it should be pretty fast imo.

Wdyt? you guys know better how this class plays within the hot-paths.

Reply:
You are right, there is an issue with the class API. Even with the `setPreferredReusableBufferType` the API is confusing as it only applies to the `ThreadLocal` logic and not to the buffer pool. The calls to `createBuffer` are also performed from within the `CompressedSegment` and `EncryptedSegment` using the same type as the `preferredReusableBufferType`. So we can pass the buffer type as part of the constructor and remove the parameter from `createBuffer`. It will fix the threading issue and the API at the same time. Wdyt?

Reply:
@bereng The approach of your last commit is also fine for me. I just tend to prefer the constructor approach because it avoid a lookup and makes the code easier to read (less parameters being passed around). That being said, I am open for both solutions.

Reply:
Indeed the API for the pool could benefit from that imo as well. Adding the param to the constructor and making both the pool and the reusableBB follow that type sgtm and a better option overall. A quick look at the code seems to suggest BBType is reachable at construction time :-)

Reply:
I have pushed the latest change where preferred BBType is set at construction time.  Despite the API being much nicer imo, given we want to move away from singletons, static state/inits, etc I wonder if this is a step in the wrong direction.

Reply:
So apparently yes, it was a step in the 'wrong' direction. We have tests that apparently change config [on the fly](https://github.com/apache/cassandra/blob/trunk/test/unit/org/apache/cassandra/db/RecoveryManagerFlushedTest.java#L71). So the static init [won't play nice](https://app.circleci.com/pipelines/github/bereng/cassandra/120/workflows/7f806bb7-4339-421f-9486-ecc47d073afd/jobs/916) with that.

Hence I propose we settle on the solution with the `EnumMap` [lookup](https://github.com/apache/cassandra/commit/1e4b2f483d221bb1adce212dac4b2109ca0e9789)

wdyt?

Reply:
@blerer suggested using a simple instance variable instead of a static one. The assumption is there will always only be one `AbstractCommitLogSegmentManager` instance. This allows for the imrpoved API and for tests to change config on the fly.